File: 0048-rotate-image/0048-rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        for (int i = 1; i < matrix[0].size(); i++) {
            for (int j = 0; j < i; j++) {
                swap(matrix[i][j], matrix[j][i]);
            }
        }

        int j = 0, k = matrix[0].size() - 1;

        while (j < k) {
            for (int i = 0; i < matrix.size(); i++) {
                swap(matrix[i][j], matrix[i][k]);
            }
            j++;
            k--;
        }
    }
};

File: 0048-rotate-image/0048-rotate-image.rs
use std::mem;

impl Solution {
    pub fn rotate(matrix: &mut Vec<Vec<i32>>) {
        for i in 0..matrix.len() {
            for j in 0..i {
                let temp = mem::take(&mut matrix[i][j]);
                matrix[i][j] = mem::replace(&mut matrix[j][i], temp);
            }
        }
        for row in matrix.iter_mut() {
            row.reverse();
        }
    }
}
